extmod/modussl_mbedtls.c

```diff
@@ -366,6 +366,14 @@ STATIC mp_uint_t socket_ioctl(mp_obj_t o_in, mp_uint_t request, uintptr_t arg, i
         mbedtls_ssl_config_free(&self->conf);
         mbedtls_ctr_drbg_free(&self->ctr_drbg);
         mbedtls_entropy_free(&self->entropy);
+    } else if (request == MP_STREAM_POLL) {
+        /* For POLL_RD, first check if ssl layer has bytes available... */
+        if (arg & MP_STREAM_POLL_RD) {
```

Comment:
What about the case when `arg` has both `MP_STREAM_POLL_RD` and `MP_STREAM_POLL_WR` set (and `mbedtls_ssl_get_bytes_avail` returns >0)?  Is it fair enough to shortcut the poll test and just return `MP_STREAM_POLL_RD` immediately, like is done in this PR?  Or should it also poll the underlying stream for writability, to potentially return both RD and WR flags set?

Some badly written code might poll for both RD and WR, but only act on WR.  Such code would not work correctly with this commit here because WR will potentially never be set.

OTOH, most code (eg uasyncio) polls either RD or WR exclusively, or if it does poll both RD and WR then it acts on whichever of those is set.  And shortcutting the check is obviously faster than checking the underlying stream as well.

Reply:
I'm sure someone will write such code as some point, so we should make it work otherwise it'll lead to subtle bugs.

Since the case of polling only for RD or WR but not for both is common, I suggest doing the shorcut only if `arg == MP_STREAM_POLL_RD`.  Otherwise fall through to call the underlying socket, and "or" the result of the underlying socket with `MP_STREAM_POLL_RD` in the case mbedtls has buffered bytes ready for reading.

Reply:
Thank you, that makes sense. I updated the PR and incorporated the poll mask changes into it.

Reply:
Testing this is tricky, as the `unix` port only supports polling on file descriptors and just completely bypasses this code... I am trying to put together a "multi" test setup.

Reply:
If you can write a network multitest that would be great.  Otherwise just a normal test in `tests/net_inet` or `tests/net_hosted` would be fine.

Reply:
@dpgeorge Would it be possible / acceptable to expose a Python interface to be able to call `->ioctl` on an object implementing the stream protocol? That would allow writing a unit test for this that works on all ports.

Reply:
You should be able to call it using `select.select` and/or `select.poll`.  You can also create a Python object inheriting from `io.IOBase` and implementing `def ioctl(self, request, arg)` which can then be passed to `ssl.wrap_socket()`.

What exactly do you want to do?

Reply:
You can do that, but it is unfortunately not useful on the `unix` port because it doesn't call `ioctl(MP_STREAM_POLL)` at all (it only calls `ioctl(MP_STREAM_GET_FILENO)` and polls the underlying file descriptor directly).